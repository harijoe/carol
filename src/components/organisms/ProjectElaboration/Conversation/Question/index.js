import React, { PropTypes } from 'react'
import styled from 'styled-components'

import { Image } from 'components'

const Wrapper = styled.div`
  float: left;
  width: 100%;
  text-align: left;
`

const StyledImage = styled(Image)`
  float: left;
`

const BubbleQuestion = styled.p`
  float: left;
  padding: 5px 10px;
  background: #d8d8d8;
  border-radius: 10px;
  max-width: 50%;
  overflow-wrap: break-word;
`

BubbleQuestion.displayName = 'BubbleQuestion'

const Question = ({ question }) => {
  if (question == null) {
    return null
  }

  return (
    <Wrapper>
      <StyledImage
        alt="quotatis"
        src={require('./logo.png')}
        width="40"
      />
      <BubbleQuestion>{question}</BubbleQuestion>
    </Wrapper>
  )
}

Question.propTypes = {
  question: PropTypes.string.isRequired,
}

export default Question
