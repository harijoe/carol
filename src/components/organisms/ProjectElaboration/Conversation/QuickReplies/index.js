import React, { PropTypes } from 'react'
import styled from 'styled-components'

import { Button, List } from 'components'

const StyledList = styled(List)`
  float: right;
  text-align: right;
`
StyledList.displayName = 'List'

const Item = styled.li`
  float: left;
  list-style: none;
  margin: 0 5px;
`

const QuickReplies = ({ quick_replies, response, reply }) => {
  // eslint-disable-next-line camelcase
  const isQuickRepliesArray = Array.isArray(quick_replies)

  if (!isQuickRepliesArray || (isQuickRepliesArray && response != null)) {
    return null
  }

  const quickRepliesList = () => (
    quick_replies.map(({ title, payload }, i) => (
      <Item key={i}>
        <Button onClick={() => reply(title, payload)}>{title}</Button>
      </Item>
    ))
  )

  return (
    <StyledList>
      {quickRepliesList()}
    </StyledList>
  )
}

QuickReplies.propTypes = {
  quick_replies: PropTypes.arrayOf(
    PropTypes.shape({
      title: PropTypes.string.isRequired,
    }),
  ),
  response: PropTypes.string,
  reply: PropTypes.func,
}

export default QuickReplies
