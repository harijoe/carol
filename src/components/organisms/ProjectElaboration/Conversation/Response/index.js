import React, { PropTypes } from 'react'
import styled from 'styled-components'

const Wrapper = styled.div`
  float: right;
  text-align: right;
  padding: 5px 10px;
  background: #00f;
  border-radius: 10px;
  color: #fff;
  max-width: 50%;
  overflow-wrap: break-word;
`

const Response = ({ response }) => {
  if (response == null) {
    return null
  }

  return (
    <Wrapper>
      {response}
    </Wrapper>
  )
}

Response.propTypes = {
  response: PropTypes.string,
}

export default Response
