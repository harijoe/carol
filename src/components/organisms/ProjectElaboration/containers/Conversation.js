import React from 'react'
import { connect } from 'react-redux'
import { push } from 'react-router-redux'
import { fromProjectElaboration, fromContext } from 'store/selectors'

import {projectElaborationSetIsNewConversation} from 'store/projectElaboration/actions'
import Conversation from '../organisms/Conversation'

const ConversationContainer = props => <Conversation {...props} />

const mapStateToProps = state => ({
  isNewConversation: fromProjectElaboration.isNewConversation(state),
  locale: fromContext.getLocale(state),
})

const mapDispatchToProps = dispatch => ({
  setIsNewConversation: payload => dispatch(projectElaborationSetIsNewConversation(payload)),
  redirectTo: url => dispatch(push(url)),
})

export default connect(mapStateToProps, mapDispatchToProps)(ConversationContainer)
