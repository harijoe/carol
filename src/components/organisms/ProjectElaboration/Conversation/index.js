import React, { Component, PropTypes } from 'react'
import styled from 'styled-components'

import Question from './Question'
import Response from './Response'
import QuickReplies from './QuickReplies'

const Wrapper = styled.div`
  maxHeight: 400px;
  overflow: scroll;
  bottom: 0;
`

class Conversation extends Component {
  static propTypes = {
    conversation: PropTypes.arrayOf(
      PropTypes.shape({
        message: PropTypes.shape({
          text: PropTypes.string.isRequired,
          quick_replies: PropTypes.array,
        }).isRequired,
        response: PropTypes.string,
      }),
    ),
    reply: PropTypes.func,
  }

  componentDidMount() {
    this.scrollTop()
  }

  componentDidUpdate() {
    this.scrollTop()
  }

  scrollTop() {
    this.history.scrollTop = this.history.scrollHeight
  }

  render() {
    const { conversation, reply } = this.props

    return (
      <Wrapper innerRef={(ref) => { this.history = ref }}>
        {
          conversation.map((field, index) => (
            <div key={index}>
              <Question question={field.message.text} />
              {
                conversation.length - 1 === index ?
                  <QuickReplies
                    quick_replies={field.message.quick_replies}
                    reply={reply}
                    response={field.response}
                  />
                  : null
              }
              <Response response={field.response} />
            </div>
          ))
        }
      </Wrapper>
    )
  }
}

export default Conversation
