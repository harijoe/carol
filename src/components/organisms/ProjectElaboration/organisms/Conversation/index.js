import React, { Component } from 'react'
import PropTypes from 'prop-types'
import styled, { css } from 'styled-components'
import { theme, mapBreakpoints, breakpoint } from 'utils/style'
import injectTranslate from 'i18n/hoc/injectTranslate'

import { ProjectElaborationQuestion, RichTextContent } from 'components'
import { RenderChildrenOneByOne } from 'containers'
import Answer from './atoms/Answer'
import QuickReplies from './molecules/QuickReplies'
import Attachment from './molecules/Attachment'

const Wrapper = styled.div`
  bottom: 0;
  margin: 0 auto;
  margin-bottom: 5.6rem;
  height: calc(100vh - 18.5rem);
  width: 100%;
  max-width: 80rem;
  overflow-y: auto;
  overflow-x: hidden;

  ${mapBreakpoints(
      bp => css`
    padding-left: ${theme(`grid.gutterWidth.${bp}`, 'rem')};
    padding-right: ${theme(`grid.gutterWidth.${bp}`, 'rem')};
  `,
    )} &::-webkit-scrollbar-track {
    border-radius: 6rem;
    background-color: ${theme('colors.grayscale.lightest')};
  }

  ${breakpoint('l')`
    padding-top: ${theme('grid.gutterWidth.xl')}rem;
  `};

  &::-webkit-scrollbar {
    height: ${theme('spaces.xs')};
    width: ${theme('spaces.xs')};
    background-color: ${theme('colors.grayscale.lightest')};
  }

  &::-webkit-scrollbar-thumb {
    border-radius: 6rem;
    background-color: ${theme('colors.grayscale.light')};
  }
`

class Conversation extends Component {
  static propTypes = {
    locale: PropTypes.string,
    translate: PropTypes.func,
    activeConversation: PropTypes.arrayOf(
      PropTypes.shape({
        message: PropTypes.shape({
          text: PropTypes.string,
          quick_replies: PropTypes.array,
        }),
        answer: PropTypes.shape({
          text: PropTypes.string,
          payload: PropTypes.string,
        }),
      }),
    ),
    reply: PropTypes.func,
    redirectTo: PropTypes.func,
    setIsNewConversation: PropTypes.func,
    isNewConversation: PropTypes.bool.isRequired,
  }

  componentDidMount() {
    this.scrollTop()
  }

  componentDidUpdate() {
    this.scrollTop()
  }

  onAllChildrenRendered = () => {
    this.props.setIsNewConversation(false)
  }

  scrollTop = () => {
    this.history.scrollTop = this.history.scrollHeight
    window.scrollTo(0, document.body.scrollHeight)
  }

  render() {
    const { activeConversation, reply, locale, redirectTo, translate, isNewConversation } = this.props
    const isLastQuestion = index => activeConversation.length - 1 === index
    const introBubbles = translate('chatbot.intro').split(/(?:<br>\n){2,}/)

    return (
      <Wrapper
        innerRef={ref => {
          this.history = ref
        }}
        className="conversation"
      >
        <RenderChildrenOneByOne interval={1000} enabled={isNewConversation} onAllRendered={this.onAllChildrenRendered}>
          {activeConversation.length !== 0 && introBubbles.map(content =>
            <ProjectElaborationQuestion key={content} onAnimationEnd={this.scrollTop}>
              <RichTextContent content={content} />
            </ProjectElaborationQuestion>
          )}
          {activeConversation.map(({ message: { text, attachment, quick_replies }, answer }, index) =>
          <div key={index}>
            { // eslint-disable-next-line camelcase
              <ProjectElaborationQuestion onAnimationEnd={this.scrollTop}>
                {text != null ? text : null}
              </ProjectElaborationQuestion>}
            {isLastQuestion(index)
              ? <div>
                  <Attachment attachment={attachment != null ? attachment : null} {...{ reply, answer, locale, redirectTo }} />
                  <QuickReplies
                    // eslint-disable-next-line camelcase
                    {...{ quick_replies, reply, answer }}
                  />
                </div>
              : null}
            <Answer answer={answer != null ? answer.text : null} />
          </div>,
        )}
        </RenderChildrenOneByOne>
      </Wrapper>
    )
  }
}

export default injectTranslate(Conversation)
