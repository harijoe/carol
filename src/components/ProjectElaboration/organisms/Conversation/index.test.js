import React from 'react'
import { shallow } from 'enzyme'

import Conversation from './'

const wrap = (props = {}) => shallow(<Conversation {...props} />)

it('renders Conversation', () => {
  expect(
    wrap({
      activeConversation: [
        {
          message: {
            text: 'message',
            quick_replies: [],
            image_url: 'url',
          },
          answer: {
            text: 'response',
            payload: '6843531-515-56465',
          },
        },
      ],
    }),
  ).toMatchSnapshot()
})
