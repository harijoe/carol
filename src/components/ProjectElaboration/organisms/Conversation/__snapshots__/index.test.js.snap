// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Conversation 1`] = `
<Conversation
  activeConversation={
    Array [
      Object {
        "answer": Object {
          "payload": "6843531-515-56465",
          "text": "response",
        },
        "message": Object {
          "image_url": "url",
          "quick_replies": Array [],
          "text": "message",
        },
      },
    ]
  }
  labelWithColon={[Function]}
  rawTranslate={[Function]}
  translate={[Function]}
/>
`;
